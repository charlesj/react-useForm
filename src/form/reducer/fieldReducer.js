import { List } from 'immutable'

export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  addListItem: 'addListItem',
  removeListItem: 'removeListItem',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validationResult: 'validationResult',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  addListItem: (fieldName, fieldState) => ({ type: actionTypes.addListItem, fieldName, payload: fieldState }),
  removeListItem: (fieldName, index) => ({ type: actionTypes.removeListItem, fieldName, payload: index }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
}

export const removedFieldsKey = 'removedFields'
export const fieldsKey = 'fields'

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName = '', payload }) => {
  let fieldPath = [fieldName]
  if (fieldName.includes('.items.')) {
    fieldPath = fieldName.split('.')
  }

  const handlers = {
    [actionTypes.updateValue]: value =>
      state.setIn([fieldsKey, ...fieldPath, current, 'value'], value)
        .setIn([fieldsKey, ...fieldPath, current, 'pristine'], value == state.getIn([fieldsKey, fieldName, 'initial', 'value'])),

    [actionTypes.touched]: () =>
      state.setIn([fieldsKey, ...fieldPath, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) =>
      state.setIn([fieldsKey, ...fieldPath, current, 'error'], error)
        .setIn([fieldsKey, ...fieldPath, current, 'helperText'], helperText),

    [actionTypes.insertField]: fieldState =>
      state.setIn([fieldsKey, ...fieldPath], fieldState)
        .deleteIn([removedFieldsKey, state.get(removedFieldsKey, List()).indexOf(fieldName)]),

    [actionTypes.removeField]: () =>
      state.deleteIn([fieldsKey, ...fieldPath])
        .set(removedFieldsKey, state.get(removedFieldsKey, List()).push(fieldName)),

    [actionTypes.addListItem]: itemState => {
      const path = [fieldsKey, ...fieldPath, 'items']
      return state.setIn([...path, state.getIn(path, List()).size], itemState)
    },

    [actionTypes.removeListItem]: index =>
      state.deleteIn([fieldsKey, ...fieldPath, 'items', index]),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {
    return handler(payload)
  }
  return state
}
