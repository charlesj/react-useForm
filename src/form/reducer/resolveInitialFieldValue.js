import { defaultTextValue } from '../fields/useTextField'
import { defaultSelectValue } from '../fields/useSelectField'
import { defaultNumberValue } from '../fields/useNumberField'
import { defaultBooleanValue } from '../fields/useBooleanField'
import { defaultListValue } from '../fields/useListField'

const defaultValues = {
  text: defaultTextValue,
  select: defaultSelectValue,
  number: defaultNumberValue,
  boolean: defaultBooleanValue,
  list: defaultListValue,
}

export const resolveInitialFieldValue = (fieldValue, initialValue, type = 'text') => {
  if (initialValue) {
    return initialValue
  }
  if (fieldValue) {
    return fieldValue
  }
  return defaultValues[type]
}
