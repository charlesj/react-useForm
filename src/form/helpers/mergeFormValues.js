import { Map, List } from 'immutable'
import { fieldsKey, removedFieldsKey } from '../reducer/fieldReducer'

export const mergeFormValues = (state, initialValues) => {
  let values = reduceFields(state, initialValues)
  // Remove any fields that were specifically removed through actions
  values = state.get(removedFieldsKey, List()).reduce((acc, fieldName) => {
    return acc.deleteIn(fieldName.split('.'))
  }, values)

  return values
}

const reduceFields = (state, initialState) => {
  return state.get(fieldsKey, Map()).entrySeq().reduce((acc, [key, fieldData]) => {
    const items = fieldData.get('items')
    const value = !items ?
      fieldData.getIn(['current', 'value']) :
      items.map(item => {
        return reduceFields(item, Map())
      })

    return acc.setIn(key.split('.'), value)
  }, initialState)
}
