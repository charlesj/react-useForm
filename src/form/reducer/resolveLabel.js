import { formatOptionalLabel } from '../helpers/formatLabel'

export const resolveLabel = (field, options) => {
  const { optionalLabelFormatter = formatOptionalLabel } = options
  return field.optional ? optionalLabelFormatter(field.label, field.name) : field.label
}
