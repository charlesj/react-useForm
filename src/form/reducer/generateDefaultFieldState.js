import { fromJS } from 'immutable'
import { resolveInitialFieldValue } from './resolveInitialFieldValue'
import { resolveLabel } from './resolveLabel'

/**
 *
 * @param field {import("..").IFormFieldArgs}
 * @param {*} initialValues
 * @param options {import("..").IFormOptions}
 */
export const generateDefaultFieldState = (field, initialValues, options) => {
  if (!field.name) throw new Error('name is required on field')

  const label = resolveLabel(field, options)
  const value = resolveInitialFieldValue(field.value, initialValues.getIn(field.name.split('.')), field.type)

  return fromJS({
    initial: {
      type: field.type || 'text',
      value,
      optional: field.optional || false,
      label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
      value,
    },
  }).setIn(['initial', 'field'], field)
}
