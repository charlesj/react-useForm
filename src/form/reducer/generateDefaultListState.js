import { fromJS, Map } from 'immutable'
import { resolveInitialFieldValue } from './resolveInitialFieldValue'
import { resolveLabel } from './resolveLabel'
import { generateDefaultFieldState } from './generateDefaultFieldState'

/**
 *
 * @param field {import("..").IFormFieldArgs}
 * @param {*} initialValues
 * @param options {import("..").IFormOptions}
 */
export const generateDefaultListState = (field, initialValues, options) => {

  const label = resolveLabel(field, options)
  const value = resolveInitialFieldValue(field.value, initialValues.getIn(field.name.split('.')), field.type)
  const items = getCurrentValues(field.fields, value, options, field.name)

  return fromJS({
    initial: {
      type: field.type,
      optional: field.optional || false,
      label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
    },
    items,
  }).setIn(['initial', 'field'], field)
    .setIn(['initial', 'options'], options)
}

export const getCurrentValues = (fields = [], initialValues, options, listField) => {
  return initialValues.map((item, index) => {
    return getFields(fields, item, options, listField, index)
  })
}

export const getFields = (fields = [], value = Map(), options, listField, index = 0) => {
  return Map([['fields', fields.reduce((acc, field) => {
    return acc.set(field.name, generateDefaultFieldState(field, value, options))
      .setIn([field.name, 'initial', 'field'], {
        ...field,
        name: `${listField}.items.${index}.fields.${field.name}`,
      })
  }, Map())]])
}
