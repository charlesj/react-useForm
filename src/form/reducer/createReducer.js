import { fromJS, Map } from 'immutable'
import { useReducer, useMemo } from 'react'
import { fieldReducer, initState, actions, fieldsKey } from './fieldReducer'
import { generateDefaultListState } from './generateDefaultListState'
import { generateDefaultFieldState } from './generateDefaultFieldState'

export const createReducer = ({ fields, options = {}, initialValues = Map() }) => {
  const derivedState = getInitialState(fields, initialValues, options)
  const [state, dispatch] = useReducer(fieldReducer, derivedState, initState)

  const json = JSON.stringify(initialValues.toJS())
  useMemo(() => {
    setTimeout(() => {
      dispatch(actions.reset(derivedState))
    })
    return json
  }, [json])

  return [state, dispatch]
}

/**
 * @param fields {Array<import("..").IFormFieldArgs>}
 */
const getInitialState = (fields, initialValues, options) => {
  const fieldMap = fields.reduce((acc, field) => {
    const result = acc.set(field.name, getFieldState(field, initialValues, options))

    return result
  }, Map())

  return fromJS({ [fieldsKey]: fieldMap })
}

const getFieldState = (field, initialValues, options) => {
  if (field.type === 'list') {
    return generateDefaultListState(field, initialValues, options)
  }
  return generateDefaultFieldState(field, initialValues, options)
}

