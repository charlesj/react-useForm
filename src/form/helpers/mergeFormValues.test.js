import { fromJS } from 'immutable'
import { mergeFormValues } from './mergeFormValues'
import { removedFieldsKey } from '../reducer/fieldReducer'

let state, initialValues
beforeEach(() => {
  initialValues = fromJS({
    id: 'the id',
    name: 'old name',
    phone: 'old phone',
  })
  state = fromJS({
    fields: {
      name: {
        current: {
          value: 'old name',
        },
      },
      phone: {
        current: {
          value: 'new phone',
        },
      },
    },
  })
})

test('merges current form values with initialValues', () => {
  expect(mergeFormValues(state, initialValues)).toMatchSnapshot()
})

test('merges current nested form values with initialValues', () => {
  initialValues = fromJS({
    id: 'the id',
    data: {
      name: 'old name',
      phone: 'old phone',
    },
  })
  state = fromJS({
    fields: {
      'data.name': {
        current: {
          value: 'old name',
        },
      },
      'data.phone': {
        current: {
          value: 'new phone',
        },
      },
    },
  })

  expect(mergeFormValues(state, initialValues)).toMatchSnapshot()
})

test('merges current list values with initialValues', () => {
  initialValues = fromJS({
    id: 'the id',
    data: {
      name: 'old name',
      phones: [
        { number: 'old phone 1' },
        { number: 'old phone 2' },
      ],
    },
  })
  state = fromJS({
    fields: {
      'data.name': {
        current: {
          value: 'old name',
        },
      },
      'data.phones': {
        current: {
        },
        items: [
          {
            fields: {
              number: {
                current: {
                  value: 'new phone 1',
                },
              },
            },
          },
          {
            fields: {
              number: {
                current: {
                  value: 'new phone 2',
                },
              },
            },
          },
        ],
      },
    },
  })

  expect(mergeFormValues(state, initialValues)).toMatchSnapshot()
})

test('submit removes fields that were removed via action', () => {
  const newState = state.set(removedFieldsKey, fromJS(['name']))
  expect(mergeFormValues(newState, initialValues)).toMatchSnapshot()
})
