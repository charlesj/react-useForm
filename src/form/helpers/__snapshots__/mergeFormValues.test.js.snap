// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merges current form values with initialValues 1`] = `
Immutable.Map {
  "id": "the id",
  "name": "old name",
  "phone": "new phone",
}
`;

exports[`merges current list values with initialValues 1`] = `
Immutable.Map {
  "id": "the id",
  "data": Immutable.Map {
    "name": "old name",
    "phones": Immutable.List [
      Immutable.Map {
        "number": "new phone 1",
      },
      Immutable.Map {
        "number": "new phone 2",
      },
    ],
  },
}
`;

exports[`merges current nested form values with initialValues 1`] = `
Immutable.Map {
  "id": "the id",
  "data": Immutable.Map {
    "name": "old name",
    "phone": "new phone",
  },
}
`;

exports[`submit removes fields that were removed via action 1`] = `
Immutable.Map {
  "id": "the id",
  "phone": "new phone",
}
`;
